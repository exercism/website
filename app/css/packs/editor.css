@import "../minimal_tailwind.css";
@import "../ui-kit/minimal.css";

@import "../pages/editor";
@import "../modals/editor-hints";
@import "../components/chatgpt";
