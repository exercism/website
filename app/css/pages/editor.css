@import "../styles";

@keyframes progress0To100 {
    from {
        width: 0%;
    }
    to {
        width: 100%;
    }
}

#page-editor {
    @apply fixed flex flex-col;
    width: 100%;
    height: 100%;

    & .header {
        @apply bg-backgroundColorA;
        @apply border-b-1 border-borderColor6;
        @apply flex-shrink-0 flex-grow-0;

        @apply flex items-center;
        @apply py-6 px-32;

        & .close-btn,
        & .options {
            flex-basis: 0;
            flex-grow: 1;
        }

        & .close-btn {
            @apply flex;
            @apply font-mono font-medium text-textColor6 text-14;
            & .c-icon {
                height: 16px;
                width: 16px;
                @apply mr-16;
                filter: var(--textColor6-filter);
            }
        }
        & .title {
            @apply flex-grow flex justify-center;
            @apply font-mono text-15 font-medium;

            & .track {
                @apply text-textColor6;
            }

            & .divider {
                @apply text-textColor6;
                @apply px-16;
            }

            & .exercise {
                @apply text-textColor2 font-medium;
            }
        }
        & .options {
            @apply flex items-center justify-end;
        }

        & .hints-btn,
        & .keyboard-shortcuts-btn,
        & .settings-btn,
        & .more-btn {
            height: 36px;
            @apply flex justify-center items-center;
        }

        & .hints-btn {
            @apply px-6;
            & .c-icon {
                filter: var(--orange-filter);
                height: 24px;
                width: 24px;
            }
        }

        & .keyboard-shortcuts-btn {
            @apply px-6;
            @apply text-textColor6;
            & .c-icon {
                height: 24px;
                width: 24px;
                filter: var(--textColor6-filter);
            }
        }

        & .hints-btn,
        & .keyboard-shortcuts-btn,
        & .settings-btn {
            @apply mr-24;
        }

        & .settings-btn,
        & .more-btn {
            @apply text-textColor6;
            @apply px-10;
            & .c-icon {
                height: 16px;
                width: 16px;
                filter: var(--textColor6-filter);
            }
        }

        & .settings-dialog {
            @apply bg-backgroundColorA;
            @apply flex flex-col shadow-lg;
            @apply py-24 px-32 text-left;
            @apply z-dropdown;

            & .setting {
                @apply flex items-center justify-between;
                @apply mb-16;
                &:last-child {
                    @apply mb-0;
                }

                & .name {
                    @apply text-16 font-medium text-darkGray;
                    @apply mr-56;
                    width: 110px;
                }

                & .options {
                    @apply bg-borderLight rounded-100;
                    @apply flex items-center flex-grow-0;
                    box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);

                    & label {
                        @apply block relative;
                        & .label {
                            @apply px-12 py-6;
                            @apply cursor-pointer;
                            @apply border-1 border-borderLight rounded-100;
                            @apply text-textColor6 font-mono font-bold leading-140;
                        }

                        /* Keep it here so it's discoverable by touch
                         * for accessibility purposes */
                        & input {
                            @apply absolute;
                            top: 0;
                            left: 0;
                            opacity: 0;
                            &:checked + .label {
                                @apply bg-lightBlue;
                                @apply text-textColor9;
                            }

                            /* It would be nice to DRY this up with application.css
                             * but its tricky with the adjecent element. */
                            &:focus {
                                & + .label {
                                    outline: 5px auto -webkit-focus-ring-color;
                                }
                                &:not(.focus-visible) + .label {
                                    outline: none;
                                }
                            }
                        }
                    }
                }
            }
        }
        & .actions-dialog {
            min-width: 380px;
        }
    }
    /* This appears in various places in the UI */
    /* This whole thing is to get round issues with
     * disabled buttons not firing mouse events */
    & .run-tests-btn {
        @apply relative;

        & .disabled-wrapper {
            @apply hidden absolute;
            top: 0;
            left: 0;
            right: 0;
            bottom: 0;
        }
        & button[disabled] + .disabled-wrapper {
            @apply block;
            cursor: not-allowed;
        }
    }

    & .c-split-pane {
        & .--split-lhs {
            @apply flex flex-col;
            @apply bg-backgroundColorB;
            @apply overflow-hidden;
            width: calc(100% - 33%);

            & .tabs {
                @apply flex;
                & .c-tab {
                    @apply mr-0;
                }
            }

            & .c-iteration-pane {
                overflow: hidden;
            }
            & .--tab-panel {
                @apply flex-grow;
                @apply overflow-hidden;
            }
            & .editor {
                @apply flex-grow;
                @apply bg-backgroundColorA;
                overflow-y: auto;
                height: 100%;

                & .ace_content {
                    @apply font-medium;
                }

                /* These bits handle things resizing correctly */
                @apply flex;
                & .ace_editor {
                    @apply flex-grow;
                    @apply w-100;
                }
            }

            /*@apply border-t-1 border-borderColor6;*/
            & footer.lhs-footer {
                @apply flex-shrink-0 flex-grow-0;
                @apply flex items-center justify-end px-32 py-24;
                @apply bg-backgroundColorB;
                @apply relative;
                box-shadow: inset 0 24px 24px -24px var(--backgroundColorA);

                & button {
                    @apply ml-20;
                }
                & .run-tests-btn {
                    & button {
                        @apply ml-0;
                    }
                }
            }
        }
        & .--split-rhs {
            @apply bg-backgroundColorA;
            @apply flex flex-col;
            width: 33%;

            /* Put a shadow over the LHS */
            @apply shadow-xsZ1 z-shadow;

            & .tabs {
                @apply flex flex-shrink-0 flex-grow-0;
                @apply py-16 px-24;

                & .c-tab {
                    height: 36px;
                    @apply mr-12;

                    & .c-icon {
                        height: 16px;
                        width: 16px;
                    }
                }
            }

            & > .--tab-panel {
                overflow-y: scroll;
                height: 100%;
            }

            & section.instructions-pane {
                @apply pb-16 px-24;

                & > .c-textual-content {
                    & * + h3 {
                        margin-top: 0;
                    } /* Override the default */
                    & :not(h2) + h3 {
                        @apply mt-28;
                    }
                    & .content {
                        @apply mt-0;
                    }
                }
                & .instructions {
                    @apply pt-16;
                    @apply border-t-1 border-borderColor6;
                }
                & .c-details.task {
                    @apply border-t-2 border-lightGray;
                    @apply mt-24 pt-16;

                    & > .--summary {
                        @apply mb-10;
                        & > .--summary-inner {
                            @apply flex items-center;

                            & .task-marker {
                                @apply mr-20;
                                @apply py-4 px-8;
                                @apply bg-lightGold text-unnamed13 rounded-3;
                                @apply font-mono font-bold text-12 leading-120;
                                @apply flex-shrink-0 self-start;
                                margin-top: 2px;
                            }
                            & .summary-title {
                                @apply flex-grow;
                                @apply text-18 font-semibold leading-140;
                            }

                            & .--closed-icon,
                            & .--open-icon {
                                height: 24px;
                                width: 24px;

                                @apply flex flex-shrink-0 items-center justify-center;
                                @apply border-gray border-1 rounded-circle;
                                @apply text-darkGray;
                                @apply ml-16;

                                & .c-icon {
                                    height: 11px;
                                    width: 11px;
                                }
                            }
                        }
                    }
                }
            }
            & .--tab-panel.tests {
                &:not([hidden]) {
                    @apply flex;
                }
                & code {
                    @apply pb-24;
                    height: 100%;
                }
            }
            & section.results {
                & .running {
                    @apply px-24 py-24;
                    @apply flex flex-col items-center;
                    & .c-icon {
                        height: 24px;
                        width: 24px;
                        animation: spin 3s linear infinite;
                        filter: var(--textColor6-filter);
                        @apply mb-20;
                    }
                    & .progress {
                        @apply w-100 rounded-100;
                        @apply bg-veryLightBlue;
                        @apply flex items-center;
                        @apply mb-12;
                        height: 6px;

                        & .bar {
                            height: 6px;
                            min-width: 6px;
                            background: linear-gradient(
                                to bottom,
                                #2200ff 0%,
                                #9e00ff 100%
                            );

                            @apply rounded-100;

                            animation: progress0To100 normal forwards;
                            animation-timing-function: cubic-bezier(
                                0.82,
                                0.27,
                                0.05,
                                0.92
                            );
                            animation-iteration-count: 1;
                            animation-duration: 3s;
                        }
                    }
                    & p {
                        @apply text-14 leading-170 font-mono font-medium;
                        @apply text-textColor6;
                        @apply mb-16;
                        & strong {
                            @apply font-semibold;
                            @apply text-textColor2;
                        }
                    }
                }
            }
            & section.run-tests-prompt {
                @apply flex flex-col items-center;
                @apply pt-32 pb-8 px-8 mb-24;

                & .c-icon {
                    height: 48px;
                    width: 48px;
                    @apply mb-16;
                }

                & h2 {
                    @apply flex items-center;
                    @apply mb-12;
                    &,
                    & button {
                        @apply text-h6 text-textColor6;
                    }
                    & button {
                        @apply py-0 px-0 mr-4;
                        & :first-child {
                            @apply mr-4;
                        }
                        & .--kb {
                            padding-top: 3px;
                            padding-bottom: 2px;
                        }
                    }
                }
                & p {
                    @apply text-15 leading-160 text-textColor6;
                    @apply text-center;
                    max-width: 450px;
                }
            }
        }
    }
}

#page-editor {
    & .cm-editor {
        width: 100%;

        &.cm-focused,
        & .focus-visible {
            /* TODO: Find a way to set outline for a11y only */
            outline: none;
        }
        & .cm-panels-top {
            @apply border-0;
        }

        & .cm-scroller {
            @apply text-15 font-mono leading-160;
        }
        & .cm-gutters {
            @apply bg-backgroundColorB;
            @apply border-borderColor7;

            & .cm-lineNumbers .cm-gutterElement {
                @apply text-center text-textColor7;
                @apply px-16;
            }
            & .cm-foldGutter .cm-gutterElement {
                @apply pr-4;
                & span {
                    vertical-align: top;
                    line-height: 13px;
                    &[title="Unfold line"] {
                        line-height: 20px;
                    }
                }
            }
        }
        & .cm-content {
            & .cm-line {
                @apply pl-8 pr-8;
            }
        }
    }
}

/* General overrides */
/* #page-editor { */
/*     & .ace_editor { */
/*         line-height: 170%; */
/*     } */
/*     & .ace_gutter, */
/*     & .ace_gutter-layer { */
/*         width: 70px !important; */

/*         & .ace_gutter-cell { */
/*             padding-left: 0px !important; */
/*             padding-right: 16px !important; */
/*             text-align: center; */
/*         } */
/*     } */
/*     & .ace_scroller { */
/*         left: 70px !important; */
/*     } */
/*     & .ace_scroller.ace_scroll-left { */
/*         box-shadow: 17px 2px 16px -16px rgba(79, 114, 205, 0.3) inset; */
/*     } */

/*     /1* Stop brackets from filling the full line height *1/ */
/*     & .ace_bracket { */
/*         max-height: 20px; */
/*         margin-top: 3px; */
/*     } */
/*     & .ace_dialog.ace_dialog-bottom { */
/*         @apply border-t-1 border-borderColor6; */
/*         @apply bg-backgroundColorB; */
/*         @apply pl-16; */
/*         &, */
/*         & input { */
/*             @apply text-15 font-semibold font-mono; */
/*             @apply text-darkGreen; */
/*         } */
/*         & input { */
/*             @apply w-100; */
/*             background: transparent; */
/*             @apply pl-4; */
/*             &:focus { */
/*                 outline: none !important; */
/*                 border: none !important; */
/*                 box-shadow: none !important; */
/*             } */
/*         } */
/*     } */
/* } */

/* /1* Light theme overrides *1/ */
/* #page-editor .ace-tm { */
/*     & .ace_gutter { */
/*         @apply bg-backgroundColorA; */
/*     } */
/*     & .ace_gutter-cell { */
/*         @apply text-textColor7; */
/*     } */
/*     /1* Same as the content active line *1/ */
/*     & .ace_gutter-cell.ace_gutter-active-line { */
/*         background: rgba(0, 0, 0, 0.07) !important; */
/*     } */
/* } */

/* #page-editor .ace-cobalt { */
/*     & .ace_gutter, */
/*     & .ace_scroller { */
/*         @apply bg-backgroundColorA; */
/*     } */
/* } */

.c-modal.modal-editor-hints {
    & .--modal-content {
        max-width: 1100px;
    }
}
