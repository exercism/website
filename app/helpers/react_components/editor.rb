module ReactComponents
  class Editor < ReactComponent
    initialize_with :solution

    def to_s
      super(
        "editor",
        {
          default_submissions: submissions,
          default_files: SerializeEditorFiles.(solution.files_for_editor),
          default_settings: {
            tab_size: track.indent_size,
            use_soft_tabs: track.indent_style == :space
          },
          autosave: {
            key: solution.uuid,
            save_interval: 500
          },
          panels: {
            instructions: {
              introduction:,
              assignment: SerializeExerciseAssignment.(solution),
              debugging_instructions:
            },
            tests: solution.exercise.practice_exercise? ? {
              tests: solution.tests,
              highlightjs_language: track.highlightjs_language
            } : nil,
            results: {
              test_runner: {
                average_test_duration: track.average_test_duration
              }
            }
          },
          track: {
            title: track.title,
            slug: track.slug
          },
          exercise: {
            title: solution.exercise.title,
            slug: solution.exercise.slug
          },
          links: {
            run_tests: Exercism::Routes.api_solution_submissions_url(solution.uuid),
            back: Exercism::Routes.track_exercise_path(track, solution.exercise)
          }
        }
      )
    end

    private
    def submissions
      submission = solution.latest_submission

      submission ? [SerializeSubmission.(submission)] : []
    end

    memoize
    def introduction
      Markdown::Parse.(solution.introduction)
    end

    memoize
    def instructions
      Markdown::Parse.(solution.instructions)
    end

    memoize
    def debugging_instructions
      return if track.debugging_instructions.blank?

      Markdown::Parse.(track.debugging_instructions)
    end

    memoize
    def track
      solution.track
    end
  end
end
