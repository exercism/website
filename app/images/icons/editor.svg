<svg id="editor" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
  <defs>
    <style>
      .a {
        fill:none;
        stroke:currentColor;
        stroke-linecap:round;
        stroke-linejoin:round;
        stroke-width:1.5px;
      }
    </style>
  </defs>
  <title>file-code-1</title>
  <path class="a" d="M22.5,21.753a1.5,1.5,0,0,1-1.5,1.5H3a1.5,1.5,0,0,1-1.5-1.5V2.253A1.5,1.5,0,0,1,3,.753H18.045a1.5,1.5,0,0,1,1.048.427l2.955,2.882A1.5,1.5,0,0,1,22.5,5.136Z"/>
  <path class="a" d="M10.5,8.253H9.75a1.409,1.409,0,0,0-1.5,1.294v1.247a1.256,1.256,0,0,1-.668,1.076L6,12.78l1.582.909a1.258,1.258,0,0,1,.668,1.076v1.248a1.408,1.408,0,0,0,1.5,1.293h.75"/>
  <path class="a" d="M13.5,8.253h.75a1.408,1.408,0,0,1,1.5,1.294v1.247a1.256,1.256,0,0,0,.668,1.076L18,12.78l-1.582.909a1.258,1.258,0,0,0-.668,1.076v1.248a1.408,1.408,0,0,1-1.5,1.293H13.5"/>
</svg>