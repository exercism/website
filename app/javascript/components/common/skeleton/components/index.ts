export { SkeletonLoader as Wrapper } from './SkeletonLoader'
export { SkeletonLine as Line } from './SkeletonLine'
export { SkeletonShape as Shape } from './SkeletonShape'
export { SkeletonText as Text } from './SkeletonText'
