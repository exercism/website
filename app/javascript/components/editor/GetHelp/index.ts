export { GetHelpTab } from './GetHelpTab'
export { GetHelpPanel } from './GetHelpPanel'
