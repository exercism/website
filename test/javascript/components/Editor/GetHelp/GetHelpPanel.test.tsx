import React from 'react'
import { render, RenderResult, screen } from '@testing-library/react'
import { GetHelpPanel } from '@/components/editor/GetHelp'
import { MOCK_DATA } from './mockdata'
import { TabsContext } from '@/components/Editor'

describe('GetHelpPanel tests', () => {
  let renderedComponent: RenderResult

  beforeEach(() => {
    renderedComponent = render(
      <TabsContext.Provider
        value={{
          current: 'get-help',
          switchToTab: () => null,
        }}
      >
        <GetHelpPanel
          assignment={MOCK_DATA.ASSIGNMENT}
          helpHtml={MOCK_DATA.HELP}
          links={{ discordRedirectPath: '', forumRedirectPath: '' }}
          track={{
            slug: 'ruby',
            title: 'Ruby',
            iconUrl: '',
            course: false,
            numConcepts: 0,
            numExercises: 0,
            numSolutions: 0,
            links: {
              self: '',
              exercises: '',
              concepts: '',
            },
          }}
        />
      </TabsContext.Provider>
    )
  })

  test('Component snapshot test', () => {
    const { asFragment } = renderedComponent

    expect(asFragment()).toMatchSnapshot()
  })

  test('Component renders hints, track help and community help', () => {
    const hint = screen.getByText(/Define the expected oven time in minutes/)
    const help = screen.getByText(/To get help if you're having trouble/)
    const communityHelp = screen.getByText('Community help')

    expect(hint).toBeInTheDocument()
    expect(help).toBeInTheDocument()
    expect(communityHelp).toBeInTheDocument()
  })
})
