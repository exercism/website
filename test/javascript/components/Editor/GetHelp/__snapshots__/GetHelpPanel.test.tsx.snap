// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetHelpPanel tests Component snapshot test 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tab-get-help"
    class="--tab-panel undefined"
    id="panel-get-help"
    role="tabpanel"
    style="display: none;"
    tabindex="0"
  >
    <div
      class="pb-12 flex flex-col"
    >
      <details
        class="c-details border-t-1 border-borderColor6 py-16 px-24 hints"
        open=""
      >
        <summary
          class="flex items-center"
        >
          <img
            alt=""
            class="c-icon mr-16 filter-textColor6"
            height="24"
            src="undefined/assets/undefined"
            width="24"
          />
          <div
            class="flex items-center justify-between w-100"
          >
            <span
              class="text-h4"
            >
              Hints and Tips
            </span>
            <span
              class="--closed-icon"
            >
              <img
                alt=""
                class="c-icon filter-textColor6"
                height="12"
                src="undefined/assets/undefined"
                width="12"
              />
            </span>
            <span
              class="--open-icon"
            >
              <img
                alt=""
                class="c-icon filter-textColor6"
                height="12"
                src="undefined/assets/undefined"
                width="12"
              />
            </span>
          </div>
        </summary>
        <div
          class="pt-8 flex flex-col gap-2"
        >
          <p
            class="text-p-base text-color-2 mb-8"
          >
            Stuck? These hints will give you nudges to the right direction to get you unblocked. Use them wisely though - remember that wrestling with a problem is where the learning occurs!
          </p>
          <p
            class="text-p-base text-color-2"
          >
            Click a heading to expand the hints:
          </p>
        </div>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              Task 1. Define the expected oven time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://www.rubyguides.com/2017/07/ruby-constants/"
                  rel="noreferrer"
                  target="_blank"
                >
                  constant
                </a>
                 that should contain the

                <a
                  href="https://ruby-doc.org/core-2.7.0/Integer.html"
                  rel="noreferrer"
                  target="_blank"
                >
                  integer
                </a>
                 value specified in the recipe.
              </li>
              

              <li>
                The 
                <code>
                  ::
                </code>
                 used in 
                <code>
                  Lasagna::EXPECTED_MINUTES_IN_OVEN
                </code>
                 means that 
                <code>
                  EXPECTED_MINUTES_IN_OVEN
                </code>
                 needs to be defined
within the 
                <code>
                  Lasagna
                </code>
                 class.
              </li>
              

            </ul>
            

          </div>
        </details>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              Task 2. Calculate the remaining oven time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://launchschool.com/books/ruby/read/methods"
                  rel="noreferrer"
                  target="_blank"
                >
                  method
                </a>
                 with a single parameter for the
actual time so far.
              </li>
              

              <li>
                You can 
                <a
                  href="https://www.freecodecamp.org/news/idiomatic-ruby-writing-beautiful-code-6845c830c664/#implicit-return"
                  rel="noreferrer"
                  target="_blank"
                >
                  implicitly return an
integer
                </a>
                 from
the method.
              </li>
              

              <li>
                You can use the 
                <a
                  href="https://www.w3resource.com/ruby/ruby-arithmetic-operators.php"
                  rel="noreferrer"
                  target="_blank"
                >
                  mathematical operator for
subtraction
                </a>
                 to subtract values.
              </li>
              

            </ul>
            

          </div>
        </details>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              Task 3. Calculate the preparation time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://launchschool.com/books/ruby/read/methods"
                  rel="noreferrer"
                  target="_blank"
                >
                  method
                </a>
                 with a single parameter for the
number of layers.
              </li>
              

              <li>
                You can 
                <a
                  href="https://www.freecodecamp.org/news/idiomatic-ruby-writing-beautiful-code-6845c830c664/#implicit-return"
                  rel="noreferrer"
                  target="_blank"
                >
                  implicitly return an
integer
                </a>
                 from
the method.
              </li>
              

              <li>
                You can use the 
                <a
                  href="https://www.w3resource.com/ruby/ruby-arithmetic-operators.php"
                  rel="noreferrer"
                  target="_blank"
                >
                  mathematical operator for
multiplication
                </a>
                 to multiply values.
              </li>
              

              <li>
                You could define an extra constant for the time in minutes per layer, or use a "magic number" in the code.
              </li>
              

            </ul>
            

          </div>
        </details>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              Task 4. Calculate the total working time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://launchschool.com/books/ruby/read/methods"
                  rel="noreferrer"
                  target="_blank"
                >
                  method
                </a>
                 with two named parameters:

                <code>
                  number_of_layers
                </code>
                 and 
                <code>
                  actual_minutes_in_oven
                </code>
                .
              </li>
              

              <li>
                You can 
                <a
                  href="https://www.freecodecamp.org/news/idiomatic-ruby-writing-beautiful-code-6845c830c664/#implicit-return"
                  rel="noreferrer"
                  target="_blank"
                >
                  implicitly return an
integer
                </a>
                 from
the method.
              </li>
              

              <li>
                You can 
                <a
                  href="http://ruby-for-beginners.rubymonstas.org/objects/calling.html"
                  rel="noreferrer"
                  target="_blank"
                >
                  invoke
                </a>
                 one of the other methods you've
defined previously.
              </li>
              

              <li>
                You can use the 
                <a
                  href="https://www.w3resource.com/ruby/ruby-arithmetic-operators.php"
                  rel="noreferrer"
                  target="_blank"
                >
                  mathematical operator for addition
                </a>
                
to add values.
              </li>
              

            </ul>
            

          </div>
        </details>
      </details>
      <details
        class="c-details border-t-1 border-borderColor6 py-16 px-24 "
        open=""
      >
        <summary
          class="flex items-center"
        >
          <img
            alt=""
            class="c-icon mr-16 filter-textColor6"
            height="24"
            src="undefined/assets/undefined"
            width="24"
          />
          <div
            class="flex items-center justify-between w-100"
          >
            <span
              class="text-h4"
            >
              Community help
            </span>
            <span
              class="--closed-icon"
            >
              <img
                alt=""
                class="c-icon filter-textColor6"
                height="12"
                src="undefined/assets/undefined"
                width="12"
              />
            </span>
            <span
              class="--open-icon"
            >
              <img
                alt=""
                class="c-icon filter-textColor6"
                height="12"
                src="undefined/assets/undefined"
                width="12"
              />
            </span>
          </div>
        </summary>
        <div
          class="pt-8 flex flex-col gap-2"
        >
          <p
            class="text-p-base text-color-2 mb-8"
          >
            Don't struggle on alone! Our community is always here to help.
          </p>
          <p
            class="text-p-base text-color-2 mb-8"
          >
            Although you can't submit a mentoring requests until you get the tests passing, you can ask for help on Discord or our Forum. Use the links below to get started:
          </p>
          <a
            class="px-12 py-10 flex gap-16 items-center rounded-8 hover:bg-backgroundColorD"
            href=""
            rel="noreferrer"
            target="_blank"
          >
            <img
              alt=""
              class="c-icon"
              height="32"
              src="undefined/assets/undefined"
              width="32"
            />
            <div>
              <h6
                class="text-h6 text-textColor1 mb-2"
              >
                Discord
              </h6>
              <p
                class="text-p-base text-textColor5"
              >
                Get real time help in the #support channel on our Discord server.
              </p>
            </div>
          </a>
          <a
            class="px-12 py-10 flex gap-16 items-center rounded-8 hover:bg-backgroundColorD"
            href=""
            rel="noreferrer"
            target="_blank"
          >
            <img
              alt=""
              class="c-icon"
              height="32"
              src="undefined/assets/undefined"
              width="32"
            />
            <div>
              <h6
                class="text-h6 text-textColor1 mb-2"
              >
                Forum
              </h6>
              <p
                class="text-p-base text-textColor5"
              >
                Dig deeper into topics and ideas on our forum.
              </p>
            </div>
          </a>
        </div>
      </details>
      <details
        class="c-details border-t-1 border-borderColor6 py-16 px-24 "
        open=""
      >
        <summary
          class="flex items-center"
        >
          <img
            alt="icon for Ruby track"
            class="c-icon c-track-icon w-[24px] h-[24px] mr-16"
            src=""
          />
          <div
            class="flex items-center justify-between w-100"
          >
            <span
              class="text-h4"
            >
              Ruby help
            </span>
            <span
              class="--closed-icon"
            >
              <img
                alt=""
                class="c-icon filter-textColor6"
                height="12"
                src="undefined/assets/undefined"
                width="12"
              />
            </span>
            <span
              class="--open-icon"
            >
              <img
                alt=""
                class="c-icon filter-textColor6"
                height="12"
                src="undefined/assets/undefined"
                width="12"
              />
            </span>
          </div>
        </summary>
        <div
          class="c-textual-content --base pt-16"
        >
          <p>
            To get help if you're having trouble, you can use one of the following resources:
          </p>
          
  
          <ul>
            
  
            <li>
              <a
                href="http://ruby-doc.org/"
                rel="noreferrer"
                target="_blank"
              >
                Ruby Documentation
              </a>
            </li>
            
  
            <li>
              <a
                href="http://stackoverflow.com/questions/tagged/ruby"
                rel="noreferrer"
                target="_blank"
              >
                StackOverflow
              </a>
            </li>
            
  
            <li>
              
  
              <a
                href="https://www.reddit.com/r/ruby"
                rel="noreferrer"
                target="_blank"
              >
                /r/ruby
              </a>
               is the Ruby subreddit.
            </li>
            
  
          </ul>
          
  
          <p>
            Should those resources not suffice, check the following pages:
          </p>
          
  
          <ul>
            
  
            <li>
              The 
              <a
                href="https://exercism.org/docs/tracks/ruby"
              >
                Ruby track's documentation
              </a>
              
  
            </li>
            
  
            <li>
              The 
              <a
                href="https://forum.exercism.org/c/programming/ruby"
                rel="noreferrer"
                target="_blank"
              >
                Ruby track's programming category on the forum
              </a>
              
  
            </li>
            
  
            <li>
              <a
                href="https://forum.exercism.org/c/programming/5"
                rel="noreferrer"
                target="_blank"
              >
                Exercism's programming category on the forum
              </a>
            </li>
            
  
            <li>
              The 
              <a
                href="https://exercism.org/docs/using/faqs"
              >
                Frequently Asked Questions
              </a>
              
  
            </li>
            
  
          </ul>
        </div>
      </details>
    </div>
  </div>
</DocumentFragment>
`;
